zebra-consensus/src/checkpoint.rs

```diff
@@ -0,0 +1,158 @@
+//! Checkpoint-based block verification for Zebra.
+//!
+//! Checkpoint-based verification uses a list of checkpoint hashes to speed up the
+//! initial chain sync for Zebra. This list is distributed with Zebra.
+//!
+//! The CheckpointVerifier compares each block's `BlockHeaderHash` against the known
+//! checkpoint hashes. If it matches, then the block is verified, and added to the
+//! `ZebraState`. Otherwise, if the block's height is lower than the maximum checkpoint
+//! height, the block awaits the verification of its child block.
+//!
+//! Verification is provided via a `tower::Service`, to support backpressure and batch
+//! verification.
+
+use futures_util::FutureExt;
+use std::{
+    collections::HashMap,
+    error,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service};
+
+use zebra_chain::block::{Block, BlockHeaderHash};
+use zebra_chain::types::BlockHeight;
+
+struct CheckpointVerifier<S> {
+    /// The underlying `ZebraState`.
+    state_service: S,
+
+    /// Each checkpoint consists of a coinbase height and block header hash.
+    ///
+    /// Checkpoints should be chosen to avoid forks or chain reorganizations,
+    /// which only happen in the last few hundred blocks in the chain.
+    /// (zcashd allows chain reorganizations up to 99 blocks, and prunes
+    /// orphaned side-chains after 288 blocks.)
+    checkpoint_list: Option<HashMap<BlockHeight, BlockHeaderHash>>,
```

Comment:
If the `Option` is used only to indicate the presence or absence of a checkpoint list below, is there a reason not to store a `HashMap` directly and use [`is_empty()`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.is_empty)?

Reply:
Yes, I also had this thought.

Reply:
Fixed in 7876e827f0a723651bda506c44ddd4ab7530ccf2.